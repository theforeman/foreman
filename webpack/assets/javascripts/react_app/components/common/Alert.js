import React from 'react';
import Icon from './Icon';
import { ALERT_CSS } from '../../constants';

const Alert = (props) => {
  let classNames = ALERT_CSS[props.type];

  if (props.css) {
    classNames += ' ' + props.css;
  }

  if (props.dismissable) {
    classNames += ' ' + ALERT_CSS.dismissable;
  }

  const closeBtn = (
    <button className="close" data-dismiss="alert" aria-hidden="true">
      <Icon type="close" />
    </button>
  );

  return (
    <div className={classNames}>
      {props.dismissable && closeBtn}
      {props.children}
    </div>
  );
};

Alert.defaultProps = { dismissable: false, css: null, type: 'info' };
export default Alert;
