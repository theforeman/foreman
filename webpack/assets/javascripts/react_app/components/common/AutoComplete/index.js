import React, { Component } from 'react';
import Downshift from 'downshift';
import { Dropdown, MenuItem, InputGroup, FormControl, Button } from 'patternfly-react';
import PropTypes from 'prop-types';

export const renderItems = ({
  items, highlightedIndex, selectedItem, getItemProps,
}) => (
  <Dropdown.Menu
    style={{
      display: 'block',
      position: 'absolute',
      left: 'inherit',
      marginTop: 0,
      top: 'auto',
    }}
  >
    {items.map((item, index) => (
      <MenuItem
        {...getItemProps({
          index,
          item,
          active: highlightedIndex === index,
          onClick: (e) => {
            // At this point the event.defaultPrevented
            // is already set to true by react-bootstrap
            // MenuItem. We need to set it back to false
            // So downshift will execute it's own handler
            e.defaultPrevented = false;
          },
        })}
        key={item}
      >
        {item}
      </MenuItem>
    ))}
  </Dropdown.Menu>
);

class AutoComplete extends Component {
  constructor(props) {
    super(props);

    this.state = {
      inputValue: '',
    };
  }

  render() {
    const {
      onSearch, labelText, onInputUpdate, items, actionText, ...rest
    } = this.props;

    return (
      <Downshift
        onStateChange={({ inputValue }) => {
          if (typeof inputValue === 'string') {
            onInputUpdate(inputValue);
            this.setState({ inputValue });
          }
        }}
        defaultHighlightedIndex={0}
        selectedItem={this.state.inputValue}
        {...rest}
      >
        {({
          getInputProps,
          getItemProps,
          getLabelProps,
          isOpen,
          inputValue,
          highlightedIndex,
          getRootProps,
          selectedItem,
        }) => (
          <div>
            {labelText && <label {...getLabelProps()}>{labelText}</label>}
            <InputGroup>
              <FormControl type="text" {...getInputProps()} />
              <InputGroup.Button>
                <Button onClick={() => onSearch(inputValue)}>{actionText}</Button>
              </InputGroup.Button>
            </InputGroup>

            {isOpen && items.length
              ? renderItems({
                  items,
                  highlightedIndex,
                  selectedItem,
                  getItemProps,
                })
              : null}
          </div>
        )}
      </Downshift>
    );
  }
}

AutoComplete.propTypes = {
  items: PropTypes.arrayOf(PropTypes.string).isRequired,
  onInputUpdate: PropTypes.func.isRequired,
  onSearch: PropTypes.func.isRequired,
  labelText: PropTypes.string,
  actionText: PropTypes.string,
};

AutoComplete.defaultProps = {
  labelText: '',
  actionText: 'Search',
};

export default AutoComplete;
