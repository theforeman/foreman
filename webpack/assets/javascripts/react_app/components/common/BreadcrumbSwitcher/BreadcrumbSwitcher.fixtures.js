export const ResourcesProps = {
  resources: [
    { caption: 'Host 1', url: '#' },
    { caption: 'Host 2', url: '#' },
    { caption: 'Host 3 with a very long name', url: '#' },
    { caption: 'Host 4', url: undefined, onClick: jest.fn() },
    { caption: 'Host 5', url: '#', onClick: undefined },
  ],
};
