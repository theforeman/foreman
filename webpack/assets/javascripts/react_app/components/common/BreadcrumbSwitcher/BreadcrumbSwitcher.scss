@import '../../../common/colors';

.scrollable-list {
  max-height: $switcher-max-height;
  overflow-y: auto;
  max-width: $switcher-max-width;
}

.no-border {
  border: none;
}
