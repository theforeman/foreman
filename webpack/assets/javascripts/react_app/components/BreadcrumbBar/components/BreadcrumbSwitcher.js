import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { Overlay } from 'react-bootstrap';

import { noop } from '../../../common/helpers';

import BreadcrumbSwitcherPopover from './BreadcrumbSwitcherPopover';
import BreadcrumbSwitcherToggler from './BreadcrumbSwitcherToggler';

class BreadcrumbSwitcher extends React.Component {
  componentDidUpdate(prevProps) {
    const { open, onOpen } = this.props;

    if (open === true && prevProps.open === false) {
      onOpen();
    }
  }

  render() {
    const {
      open,
      currentPage,
      totalPages,
      isLoadingResources,
      hasError,
      resources,
      onTogglerClick,
      onHide,
      onNextPageClick,
      onPrevPageClick,
      onSearchChange,
      searchValue,
      onSearchClear,
      searchDebounceTimeout,
      onResourceClick,
    } = this.props;

    return (
      <div className="breadcrumb-switcher" style={{ position: 'relative' }}>
        <BreadcrumbSwitcherToggler
          id="switcher"
          onClick={() => onTogglerClick()}
          ref={ref => {
            this.togglerRef = ref;
          }}
        />

        <Overlay
          rootClose
          show={open}
          container={this}
          placement="bottom"
          onHide={onHide}
          // TODO: try to remove the `ReactDOM.findDOMNode`
          // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-find-dom-node.md
          // react-bootstrap still have it in their docs: https://react-bootstrap.github.io/components/overlays/
          // eslint-disable-next-line react/no-find-dom-node
          target={() => ReactDOM.findDOMNode(this.togglerRef)}
        >
          <BreadcrumbSwitcherPopover
            id="breadcrumb-switcher-popover"
            loading={isLoadingResources}
            hasError={hasError}
            onSearchChange={onSearchChange}
            resources={resources}
            onNextPageClick={onNextPageClick}
            onPrevPageClick={onPrevPageClick}
            currentPage={currentPage}
            totalPages={totalPages}
            searchValue={searchValue}
            onSearchClear={onSearchClear}
            searchDebounceTimeout={searchDebounceTimeout}
            onResourceClick={onResourceClick}
          />
        </Overlay>
      </div>
    );
  }
}

BreadcrumbSwitcher.propTypes = {
  searchValue: PropTypes.string,
  open: PropTypes.bool,
  searchDebounceTimeout: PropTypes.number,
  currentPage: PropTypes.number,
  totalPages: PropTypes.number,
  isLoadingResources: PropTypes.bool,
  hasError: PropTypes.bool,
  resources: BreadcrumbSwitcherPopover.propTypes.resources,
  onTogglerClick: PropTypes.func,
  onHide: PropTypes.func,
  onOpen: PropTypes.func,
  onPrevPageClick: PropTypes.func,
  onNextPageClick: PropTypes.func,
  onResourceClick: PropTypes.func,
  onSearchChange: PropTypes.func,
  onSearchClear: PropTypes.func,
};

BreadcrumbSwitcher.defaultProps = {
  searchValue: '',
  open: false,
  searchDebounceTimeout: 300,
  currentPage: 1,
  totalPages: 1,
  isLoadingResources: false,
  hasError: false,
  resources: [],
  onTogglerClick: noop,
  onHide: noop,
  onOpen: noop,
  onResourceClick: noop,
  onPrevPageClick: noop,
  onNextPageClick: noop,
  onSearchChange: noop,
  onSearchClear: noop,
};

export default BreadcrumbSwitcher;
