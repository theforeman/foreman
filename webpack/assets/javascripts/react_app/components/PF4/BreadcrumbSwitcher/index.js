import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import {
  Popover,
  Button,
  Menu,
  SearchInput,
  Divider,
  MenuContent,
  MenuFooter,
  MenuList,
  MenuSearch,
  MenuItem,
  Tooltip,
  Spinner,
  MenuSearchInput,
} from '@patternfly/react-core';
import { ExchangeAltIcon } from '@patternfly/react-icons';
import { translate as __ } from '../../../common/I18n';
import Pagination from '../../Pagination';
import { removeLastSlashFromPath } from '../../../common/helpers';
import './breadcrumb-switcher.scss';

const BreadcrumbSwitcher = ({
  isOpen,
  isLoading,
  hasError,
  items,
  currentPage,
  total,
  openSwitcher,
  onHide,
  onOpen,
  onSearchChange,
  onSetPage,
  onPerPageSelect,
  perPage,
  searchValue,
  onSearchClear,
  searchDebounceTimeout,
  onResourceClick,
}) => {
  let menuListItems = [];
  const isActive = (href, id, name) => {
    const hrefWithName = href
      ? removeLastSlashFromPath(href.replace(id, name))
      : href;
    const getCurrentPath = () =>
      removeLastSlashFromPath(window.location.pathname);
    return (
      href === window.location.pathname || getCurrentPath() === hrefWithName
    );
  };

  if (isLoading) {
    menuListItems = (
      <MenuItem
        isDisabled
        key="loading"
        className="breadcrumb-switcher-spinner"
        aria-label="loading spinner"
      >
        <span>
          <Spinner />
        </span>
      </MenuItem>
    );
  } else if (hasError) {
    menuListItems = (
      <MenuItem isDisabled key="error">
        {__('An error occurred.')}
      </MenuItem>
    );
  } else if (items.length === 0) {
    menuListItems = (
      <MenuItem isDisabled key="no result">
        {__('No results found')}
      </MenuItem>
    );
  } else {
    menuListItems = items.map(({ href, id, name }) => (
      <MenuItem
        id={`${id}-${__(name)}`}
        key={id}
        itemId={id}
        to={href}
        onClick={e => onResourceClick(e, href)}
        isSelected={isActive(href, id, name)}
      >
        <Tooltip content={__(name)}>
          <span>{__(name)}</span>
        </Tooltip>
      </MenuItem>
    ));
  }
  useEffect(() => {
    if (typingTimeout) {
      return () => clearTimeout(typingTimeout);
    }
    return undefined;
  }, [typingTimeout]);
  const [typingTimeout, setTypingTimeout] = useState(null);
  const autoSearch = searchTerm => {
    if (typingTimeout) clearTimeout(typingTimeout);
    setTypingTimeout(
      setTimeout(() => onSearchChange(searchTerm), searchDebounceTimeout)
    );
  };
  return (
    <div className="pf4-breadcrumb-switcher">
      <Popover
        className="pf4-breadcrumb-switcher-popover"
        isVisible={!!isOpen}
        shouldOpen={() => {
          openSwitcher();
          onOpen();
        }}
        shouldClose={onHide}
        hasNoPadding
        showClose={false}
        appendTo={() => document.querySelector('.pf4-breadcrumb-switcher')}
        position="bottom"
        enableFlip={false}
        bodyContent={
          <div>
            <Menu ouiaId="breadcrumb-menu" isScrollable>
              <MenuSearch>
                <MenuSearchInput>
                  <SearchInput
                    value={searchValue}
                    aria-label="Filter breadcrumb items"
                    type="search"
                    onChange={(_e, value) => {
                      autoSearch(value || '');
                    }}
                    onClear={onSearchClear}
                  />
                </MenuSearchInput>
              </MenuSearch>
              <Divider />
              <MenuContent>
                <MenuList>{menuListItems}</MenuList>
              </MenuContent>
              <MenuFooter>
                <Pagination
                  ouiaId="breadcrumb-pagination"
                  itemCount={total}
                  page={currentPage}
                  perPage={perPage}
                  onSetPage={onSetPage}
                  onPerPageSelect={onPerPageSelect}
                  updateParamsByUrl={false}
                  isCompact
                />
              </MenuFooter>
            </Menu>
          </div>
        }
      >
        <Button
          ouiaId="breadcrumb-button"
          variant="plain"
          aria-label="open breadcrumb switcher"
        >
          <ExchangeAltIcon />
        </Button>
      </Popover>
    </div>
  );
};

BreadcrumbSwitcher.propTypes = {
  onSearchChange: PropTypes.func.isRequired,
  onSetPage: PropTypes.func.isRequired,
  onPerPageSelect: PropTypes.func.isRequired,
  perPage: PropTypes.number.isRequired,
  hasError: PropTypes.bool.isRequired,
  onSearchClear: PropTypes.func.isRequired,
  onResourceClick: PropTypes.func.isRequired,
  items: PropTypes.array.isRequired,
  searchDebounceTimeout: PropTypes.number.isRequired,
  searchValue: PropTypes.string.isRequired,
  currentPage: PropTypes.number,
  total: PropTypes.number.isRequired,
  isOpen: PropTypes.bool.isRequired,
  isLoading: PropTypes.bool.isRequired,
  openSwitcher: PropTypes.func.isRequired,
  onOpen: PropTypes.func.isRequired,
  onHide: PropTypes.func.isRequired,
};

BreadcrumbSwitcher.defaultProps = {
  currentPage: 1,
};

export default BreadcrumbSwitcher;
