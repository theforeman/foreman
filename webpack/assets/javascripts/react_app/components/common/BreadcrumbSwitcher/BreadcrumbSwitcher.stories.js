import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { ResourcesProps } from './BreadcrumbSwitcher.fixtures';
import BreadcrumbSwitcher from './';

storiesOf('Breadcrumb Switcher', module).add('breadcrumbs switcher', () => (
  <BreadcrumbSwitcher
    resources={[
      ...ResourcesProps.resources,
      {
        caption: 'Item with onClick',
        url: undefined,
        onClick: action('Item has been clicked'),
      },
    ]}
  />
));
