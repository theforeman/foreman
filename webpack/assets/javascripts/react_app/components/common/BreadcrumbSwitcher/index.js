import React from 'react';
import {
  Popover,
  Button,
  Icon,
  ListGroup,
  ListGroupItem,
  OverlayTrigger,
} from 'patternfly-react';
import { ButtonToolbar } from 'react-bootstrap';
import EllipsisWithTooltip from 'react-ellipsis-with-tooltip';
import PropTypes from 'prop-types';
import './BreadcrumbSwitcher.scss';

const BreadcrumbSwitcher = ({ resources }) => {
  const popover = (
    <Popover id="popover-contained">
      <ListGroup className="scrollable-list">
        {resources.map((item) => {
          const { onClick, url, caption } = item;
          return (
            <ListGroupItem
              className="no-border"
              key={`id-${caption}`}
              href={url}
              onClick={onClick}
            >
              <EllipsisWithTooltip>{caption}</EllipsisWithTooltip>
            </ListGroupItem>
          );
        })}
      </ListGroup>
    </Popover>
  );
  return (
    <ButtonToolbar>
      <OverlayTrigger
        rootClose
        trigger="click"
        placement="bottom"
        overlay={popover}
      >
        <Button>
          <Icon type="fa" name="exchange" />
        </Button>
      </OverlayTrigger>
    </ButtonToolbar>
  );
};

BreadcrumbSwitcher.propTypes = {
  href: PropTypes.string,
  onClick: PropTypes.func,
  caption: PropTypes.string,
};

export default BreadcrumbSwitcher;
