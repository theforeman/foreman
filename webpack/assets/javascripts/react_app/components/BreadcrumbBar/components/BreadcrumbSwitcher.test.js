import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { shallowRenderComponentWithFixtures } from '../../../common/testHelpers';

import BreadcrumbSwitcher from './BreadcrumbSwitcher';
import {
  breadcrumbSwitcherLoading,
  breadcrumbSwitcherLoaded,
  breadcrumbSwitcherLoadedWithPagination,
} from '../BreadcrumbBar.fixtures';

const createStubs = () => ({
  onOpen: jest.fn(),
  onHide: jest.fn(),
  onTogglerClick: jest.fn(),
  onPrevPageClick: jest.fn(),
  onNextPageClick: jest.fn(),
});

const fixtures = {
  'render closed': { open: false, ...createStubs() },
  'render loading state': {
    open: true,
    ...breadcrumbSwitcherLoading,
    ...createStubs(),
  },
  'render resources list': {
    open: true,
    ...breadcrumbSwitcherLoaded,
    ...createStubs(),
  },
  'render resources list with pagination': {
    open: true,
    ...breadcrumbSwitcherLoadedWithPagination,
    ...createStubs(),
  },
};

describe('BreadcrumbSwitcher', () => {
  describe('rendering', () => {
    const components = shallowRenderComponentWithFixtures(
      BreadcrumbSwitcher,
      fixtures
    );

    const filterSnapshotGarbage = componentJson => {
      const componentOverlay = componentJson.children.filter(
        ({ type }) => type === 'Overlay'
      )[0];
      delete componentOverlay.props.container;
      return componentJson;
    };

    const testBreadcrumbSwitcherSnapshot = (description, component) =>
      it(description, () =>
        expect(filterSnapshotGarbage(toJson(component))).toMatchSnapshot()
      );

    components.forEach(({ description, component }) =>
      testBreadcrumbSwitcherSnapshot(description, component)
    );
  });

  describe('triggering', () => {
    it('should correctly trigger onOpen', () => {
      let openCounter = 0;
      const onOpen = jest.fn();
      const component = mount(
        <BreadcrumbSwitcher open={false} onOpen={onOpen} />
      );

      const expectOpOpenCallsToMatchCounter = () =>
        expect(onOpen.mock.calls).toHaveLength(openCounter);

      const open = () => {
        openCounter += 1;
        component.setProps({ open: true });
      };
      const close = () => component.setProps({ open: false });

      expectOpOpenCallsToMatchCounter();
      open();
      expectOpOpenCallsToMatchCounter();
      close();
      expectOpOpenCallsToMatchCounter();
      open();
      expectOpOpenCallsToMatchCounter();
    });
  });
});
