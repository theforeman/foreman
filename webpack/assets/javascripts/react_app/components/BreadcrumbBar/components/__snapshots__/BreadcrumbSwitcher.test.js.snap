// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbSwitcher rendering render closed 1`] = `
<div
  className="breadcrumb-switcher"
  style={
    Object {
      "position": "relative",
    }
  }
>
  <BreadcrumbSwitcherToggler
    id="switcher"
    onClick={[Function]}
  />
  <Overlay
    animation={[Function]}
    onHide={[Function]}
    placement="bottom"
    rootClose={true}
    show={false}
    target={[Function]}
  >
    <BreadcrumbSwitcherPopover
      currentPage={1}
      hasError={false}
      id="breadcrumb-switcher-popover"
      loading={false}
      onNextPageClick={[Function]}
      onPrevPageClick={[Function]}
      onResourceClick={[Function]}
      resources={Array []}
      searchValue=""
      totalPages={1}
    />
  </Overlay>
</div>
`;

exports[`BreadcrumbSwitcher rendering render loading state 1`] = `
<div
  className="breadcrumb-switcher"
  style={
    Object {
      "position": "relative",
    }
  }
>
  <BreadcrumbSwitcherToggler
    id="switcher"
    onClick={[Function]}
  />
  <Overlay
    animation={[Function]}
    onHide={[Function]}
    placement="bottom"
    rootClose={true}
    show={true}
    target={[Function]}
  >
    <BreadcrumbSwitcherPopover
      currentPage={1}
      hasError={false}
      id="breadcrumb-switcher-popover"
      loading={false}
      onNextPageClick={[Function]}
      onPrevPageClick={[Function]}
      onResourceClick={[Function]}
      resources={Array []}
      searchValue=""
      totalPages={1}
    />
  </Overlay>
</div>
`;

exports[`BreadcrumbSwitcher rendering render resources list 1`] = `
<div
  className="breadcrumb-switcher"
  style={
    Object {
      "position": "relative",
    }
  }
>
  <BreadcrumbSwitcherToggler
    id="switcher"
    onClick={[Function]}
  />
  <Overlay
    animation={[Function]}
    onHide={[Function]}
    placement="bottom"
    rootClose={true}
    show={true}
    target={[Function]}
  >
    <BreadcrumbSwitcherPopover
      currentPage={1}
      hasError={false}
      id="breadcrumb-switcher-popover"
      loading={false}
      onNextPageClick={[Function]}
      onPrevPageClick={[Function]}
      onResourceClick={[Function]}
      resources={
        Array [
          Object {
            "id": "1",
            "name": "Host 1",
            "url": "#",
          },
          Object {
            "id": "2",
            "name": "Host 2",
            "url": "#",
          },
          Object {
            "id": "3",
            "name": "Host 3 with a very long name",
            "url": "#",
          },
          Object {
            "id": "4",
            "name": "Host 4",
            "onClick": [Function],
            "url": undefined,
          },
          Object {
            "id": "5",
            "name": "Host 5",
            "onClick": undefined,
            "url": "#",
          },
        ]
      }
      searchValue=""
      totalPages={1}
    />
  </Overlay>
</div>
`;

exports[`BreadcrumbSwitcher rendering render resources list with pagination 1`] = `
<div
  className="breadcrumb-switcher"
  style={
    Object {
      "position": "relative",
    }
  }
>
  <BreadcrumbSwitcherToggler
    id="switcher"
    onClick={[Function]}
  />
  <Overlay
    animation={[Function]}
    onHide={[Function]}
    placement="bottom"
    rootClose={true}
    show={true}
    target={[Function]}
  >
    <BreadcrumbSwitcherPopover
      currentPage={2}
      hasError={false}
      id="breadcrumb-switcher-popover"
      loading={false}
      onNextPageClick={[Function]}
      onPrevPageClick={[Function]}
      onResourceClick={[Function]}
      resources={
        Array [
          Object {
            "id": "1",
            "name": "Host 1",
            "url": "#",
          },
          Object {
            "id": "2",
            "name": "Host 2",
            "url": "#",
          },
          Object {
            "id": "3",
            "name": "Host 3 with a very long name",
            "url": "#",
          },
          Object {
            "id": "4",
            "name": "Host 4",
            "onClick": [Function],
            "url": undefined,
          },
          Object {
            "id": "5",
            "name": "Host 5",
            "onClick": undefined,
            "url": "#",
          },
        ]
      }
      searchValue=""
      totalPages={3}
    />
  </Overlay>
</div>
`;
